### PR TITLE
fix/FORMS-994 validate forms routes params

### PR DESCRIPTION
**Pull request by WalterMoar**
# Description

James P from JEDI reported a problem when calling API routes that have both `formId` and `formVersionId` parameters. The data for the `formVersionId` will be returned even if the form version does not belong to the given `formId`. Whenever these two parameters exist on a route we need to check that form contains the form version, and return a 404 response if not. We also need to check the routes that have both `formId` and `formVersionDraftId`.

While this is a potential security problem, the exploit would involve having the `formVersionId` for someone else’s form. The large key space of UUIDs protects us here, but the problem still needs to be fixed.

### Breaking Changes:

The following routes were incorrectly returning _HTTP 422 Unprocessable Content_ when the `formVersionId` was not a valid UUID. These routes now return _HTTP 400 Bad Request_ for invalid UUIDs.
- `GET /api/v1/forms/:formId/versions/:formVersionId`
- `GET /api/v1/forms/:formId/versions/:formVersionId/fields`
- `POST /api/v1/forms/:formId/versions/:formVersionId/multiSubmission`
- `GET /api/v1/forms/:formId/versions/:formVersionId/publish`
- `GET /api/v1/forms/:formId/versions/:formVersionId/submissions`
- `POST /api/v1/forms/:formId/versions/:formVersionId/submissions`
- `GET /api/v1/forms/:formId/versions/:formVersionId/submissions/discover`

The following routes were incorrectly returning _HTTP 422 Unprocessable Content_ when the `formVersionDraftId` was not a valid UUID. These routes now return _HTTP 400 Bad Request_ for invalid UUIDs.
- `DELETE /api/v1/forms/:formId/drafts/:formVersionDraftId`
- `GET /api/v1/forms/:formId/drafts/:formVersionDraftId`
- `PUT /api/v1/forms/:formId/drafts/:formVersionDraftId`
- `POST /api/v1/forms/:formId/drafts/:formVersionDraftId/publish`

## Types of changes

Bug fix (non-breaking change which fixes an issue)

Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have checked that unit tests pass locally with my changes
- [x] I have run the npm script lint on the frontend and backend
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] I have approval from the product owner for the contribution in this pull request

### CODE REVIEW

**Comment by jimmyPasta** [comment]
Looks good thanks for the quick turn around, if you deploy this to a pr env I will verify the fix.
I will check also that you can not POST to versions for a version change or grab Submissions

**Review by jimmyPasta — APPROVED** [review]
Let me know if you deploy to a pr env first and I can do some basic testing

**Comment by WalterMoar** [comment]
> Let me know if you deploy to a pr env first and I can do some basic testing

Thank you! It's deployed to https://chefs-dev.apps.silver.devops.gov.bc.ca/pr-1275

File: app/src/forms/auth/middleware/params.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+ */
+const validateFormVersionDraftId = async (req, _res, next, formVersionDraftId) => {
+  try {
+    if (!uuid.validate(formVersionDraftId)) {
+      throw new Problem(400, {
+        detail: 'Bad formVersionDraftId',
+      });
+    }
+
+    const formVersionDraft = await formService.readDraft(formVersionDraftId);
```
Comment (WalterMoar) [inline app/src/forms/auth/middleware/params.js:45]: This service call does add another database query to these routes, and the same in `validateFormVersionId` below.

We could think about doing some some of caching of these values once they're pulled from the database, but that would make things more complicated. Plus we'd have to deal with cache invalidation and things like that. Maybe keep it simple for now and do optimization later if we need it.

Ideally the database calls should be small and quick. The database itself should be caching these things and do a better job of it than we would!

**Review by jasonchung1871 — APPROVED** [review]

File: app/src/forms/auth/middleware/params.js (hunk 1 above)
Comment (usingtechnology) [reply app/src/forms/auth/middleware/params.js:45]: Am I missing something here? (I haven't run this). For draft and form version don't you need the form id too to validate if they belong to each other?

Comment (usingtechnology) [reply app/src/forms/auth/middleware/params.js:45]: offline discussion cleared this up. early morning reading comprehension must be low.

**Review by usingtechnology — APPROVED** [review]

**Pull request merged**